test/catch.cc

```diff
@@ -0,0 +1,215 @@
+    const char* what() const noexcept override {
+      return "child exception";
+    }
+  };
+
+  auto expected = []() -> Expected::Of<int> {
+    return Unexpected(Error{});
+  };
+
+  auto e = [&]() {
+    return expected() // Throwing 'std::exception_ptr' there.
+        | Catch()
+              .raised<int>([](auto&& error) {
+                ADD_FAILURE() << "Encountered an unexpected raised 'int'";
+                return 1;
+              })
+              // Receive 'Error' type there, that had been rethrowed from
+              // 'std::exception_ptr'.
+              .raised<Error>([](auto&& error) {
+                EXPECT_EQ(std::string("child exception"), error.what());
+                return 100;
+              });
+  };
+
+  EXPECT_EQ(*e(), 100);
+}
+
+TEST(CatchTest, UnexpectedAll) {
+  auto expected = []() -> Expected::Of<int> {
+    return Unexpected(10);
+  };
+
+  auto checker = [](auto error) {
+    try {
+      std::rethrow_exception(error);
+    } catch (int x) {
+      EXPECT_EQ(x, 10);
+    } catch (...) {
+      FAIL() << "Failure on rethrowing";
+    }
+  };
+
+  auto e = [&]() {
+    return expected() // Throwing 'std::exception_ptr' there.
+        | Catch()
+              .raised<double>([](auto&& error) {
+                ADD_FAILURE() << "Encountered an unexpected raised 'double'";
+                return 1;
+              })
+              .raised<std::exception>([](auto&& error) {
+                ADD_FAILURE()
+                    << "Encountered an unexpected raised 'std::exception'";
+                return 1;
+              })
+              // Receive 'std::exception_ptr' there.
+              .all([&](auto&&... error) {
+                if constexpr (std::is_same_v<
+                                  decltype(error)...,
+                                  std::exception_ptr>) {
+                  checker(error...);
+                }
+
+                return 100;
+              });
+  };
+
+  EXPECT_EQ(*e(), 100);
+}
+
+TEST(CatchTest, NoExactHandler) {
+  auto e = []() {
+    return Just(1)
+        | Raise(std::string("error"))
+        | Catch()
+              .raised<double>([](auto&& error) {
+                ADD_FAILURE() << "Encountered an unexpected raised 'double'";
+                return 1;
+              })
+              .raised<std::exception>([](auto&& error) {
+                ADD_FAILURE()
+                    << "Encountered an unexpected raised 'std::exception'";
+                return 1;
+              });
+  };
+
+  EXPECT_THROW(*e(), std::string);
+}
+
+TEST(CatchTest, ReRaise) {
+  auto e = [&]() {
+    return Just(1)
+        | Raise(1)
+        | Catch()
+              .raised<int>([](auto&& error) {
+                return Raise(1);
+              })
+              .raised<std::exception_ptr>([](auto&& error) {
+                ADD_FAILURE()
+                    << "Encountered an unexpected raised 'std::exception_ptr'";
+                return Raise(100);
```

Comment:
To really make sure that this test is doing what we want let's make one of the handlers _not_ do a `Raise()` but instead return just an `int`. I think that will mean that we actually need to do _more_ type unification to deal with the case that we have 'Raise()`. See `eventuals/conditional.h` for how to `Unify`!